src/librustc_mir/transform/match_branches.rs

```diff
@@ -0,0 +1,102 @@
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, src: MirSource<'tcx>, body: &mut Body<'tcx>) {
+        let param_env = tcx.param_env(src.def_id());
+        let mut did_remove_blocks = false;
+        let bbs = body.basic_blocks_mut();
+        'outer: for bb_idx in bbs.indices() {
+            let (discr, val, switch_ty, targets) = match bbs[bb_idx].terminator().kind {
+                TerminatorKind::SwitchInt {
+                    discr: Operand::Move(ref place),
+                    switch_ty,
+                    ref targets,
+                    ref values,
+                    ..
+                } if targets.len() == 2 && values.len() == 1 => {
+                    (place.clone(), values[0], switch_ty, targets)
+                }
+                _ => continue,
+            };
+            let (first, rest) = if let ([first], rest) = targets.split_at(1) {
+                (*first, rest)
+            } else {
+                unreachable!();
+            };
+            let first_dest = bbs[first].terminator().kind.clone();
+            let same_destinations = rest
+                .iter()
+                .map(|target| &bbs[*target].terminator().kind)
+                .all(|t_kind| t_kind == &first_dest);
+            if !same_destinations {
+                continue;
+            }
+            let first_stmts = &bbs[first].statements;
+            for s in first_stmts.iter() {
+                match &s.kind {
+                    StatementKind::Assign(box (_, rhs)) => {
+                        if let Rvalue::Use(Operand::Constant(_)) = rhs {
+                        } else {
+                            continue 'outer;
+                        }
+                    }
+                    _ => continue 'outer,
+                }
+            }
+            for target in rest.iter() {
+                for s in bbs[*target].statements.iter() {
+                    if let StatementKind::Assign(box (ref lhs, rhs)) = &s.kind {
+                        if let Rvalue::Use(Operand::Constant(_)) = rhs {
+                            let has_matching_assn = first_stmts
+                                .iter()
+                                .find(|s| {
+                                    if let StatementKind::Assign(box (lhs_f, _)) = &s.kind {
+                                        lhs_f == lhs
+                                    } else {
+                                        false
+                                    }
+                                })
+                                .is_some();
+                            if has_matching_assn {
+                                continue;
+                            }
+                        }
+                    }
+
+                    continue 'outer;
+                }
+            }
+            let (first_block, to_add) = bbs.pick2_mut(first, bb_idx);
+            let new_stmts = first_block.statements.iter().cloned().map(|mut s| {
```

Comment:
instead of moving all the statements, you can also just change the current block's terminator into a `goto` to the `first` block and mutate the `first` block's statement. We have further optimizations that will merge the blocks if applicable.